substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/PosixJavaNetSubstitutions.java

```diff
@@ -119,84 +419,1339 @@ protected int peekData(DatagramPacket p) throws IOException {
+            len_Pointer.write(JavaNetNetUtilMD.SOCKADDR_LEN());
+            //  1762
+            //  1763          if (getsockname(fd, (struct sockaddr *)&him, &len) == -1) {
+            if (Target_os.get_sock_name(fd, him, len_Pointer) == -1) {
+                //  1764              NET_ThrowByNameWithLastError(env, JNU_JAVANETPKG "SocketException",
+                //  1765                             "Error getting socket name");
+                throw new SocketException("Error getting socket name");
+                //  1766              return NULL;
+            }
+            //  1768          iaObj = NET_SockaddrToInetAddress(env, (struct sockaddr *)&him, &port);
+            iaObj = JavaNetNetUtil.NET_SockaddrToInetAddress(him, port_Pointer);
+            //  1769
+            //  1770          return iaObj;
+            return iaObj;
+        }
+
+        //  1773      /*
+        //  1774       * Map the Java level socket option to the platform specific
+        //  1775       * level and option name.
+        //  1776       */
+        //  1777      if (NET_MapSocketOption(opt, &level, &optname)) {
+        if (JavaNetNetUtilMD.NET_MapSocketOption(opt, level_Pointer, optname_Pointer) != 0) {
+            //  1778          JNU_ThrowByName(env, JNU_JAVANETPKG "SocketException", "Invalid option");
+            //  1779          return NULL;
+            throw new SocketException("Invalid option");
+        }
+
+        //  1782      if (opt == java_net_SocketOptions_IP_MULTICAST_LOOP &&
+        //  1783          level == IPPROTO_IP) {
+        if (opt == SocketOptions.IP_MULTICAST_LOOP && level_Pointer.read() == NetinetIn.IPPROTO_IP()) {
+            //  1784          optlen = sizeof(optval.c);
+            optlen_Pointer.write(SizeOf.get(CCharPointer.class));
+            optval_Pointer = optval_c_Pointer;
+        } else {
+            //  1786          optlen = sizeof(optval.i);
+            optlen_Pointer.write(SizeOf.get(CIntPointer.class));
+            optval_Pointer = optval_i_Pointer;
+        }
+        //  1788
+        //  1789      if (NET_GetSockOpt(fd, level, optname, (void *)&optval, &optlen) < 0) {
+        if (JavaNetNetUtilMD.NET_GetSockOpt(fd, level_Pointer.read(), optname_Pointer.read(), optval_Pointer, optlen_Pointer) < 0) {
+            //  1790          NET_ThrowByNameWithLastError(env, JNU_JAVANETPKG "SocketException",
+            //  1791                           "Error getting socket option");
+            //  1792          return NULL;
+            throw new SocketException("Error getting socket option");
+        }
+
+        //  1795      switch (opt) {
+        //  1796          case java_net_SocketOptions_IP_MULTICAST_LOOP:
+        if (opt == SocketOptions.IP_MULTICAST_LOOP) {
+            //  1798              if (level == IPPROTO_IP) {
+            if (level_Pointer.read() == NetinetIn.IPPROTO_IP()) {
+                //  1799                  return createBoolean(env, (int)!optval.c);
```

Comment:
Since you need these more than once, I think you should leave them as calls to methods.  

That also separates the computing of the argument to the method (an `int`) from the implementation of the method.  I do not want to track the `!` and `== 0` here through the `!= 0` in `createBoolean`.

I would have a `createInteger` method, even though it is trivial.  Our compilers are good at inlining trivial methods.